tests/io/test_DFastMIConfigParser.py

```diff
@@ -0,0 +1,579 @@
+"""
+Tests for DFastMIConfig class
+"""
+import sys
+from mock import Mock
+from contextlib import contextmanager
+from io import StringIO
+from typing import Tuple
+import pytest
+import configparser
+
+from dfastmi.io.IBranch import IBranch
+from dfastmi.io.Reach import Reach, ReachAdvanced
+
+from dfastmi.io.DFastMIConfigParser import DFastMIConfigParser
+
+@contextmanager
+def captured_output():
+    new_out, new_err = StringIO(), StringIO()
+    old_out, old_err = sys.stdout, sys.stderr
+    try:
+        sys.stdout, sys.stderr = new_out, new_err
+        yield sys.stdout, sys.stderr
+    finally:
+        sys.stdout, sys.stderr = old_out, old_err
+
+class Test_DFastMIConfigParser():
+    """
+    Class will validate the functionality the DFastMConfigParser
+    """
+
+    _reach : Reach
+
+    @pytest.fixture
+    def setup_data(self):
+        """Setup a reach with a branch to be used in the tests"""
+        self._reach = Reach("Reach1")
+        branch = Mock(IBranch)
+        branch.name = "Branch1"
+        self._reach.parent_branch = branch
+
+    def given_simple_river_config_when_read_key_with_custom_unknown_type_then_exception_thrown(self, setup_data):
+        """
+        When we want to read a configuration value from a river configuration file but have not registered a type
+        or actually have an exception while parsing the value string to the type we want we throw an exception.
+        """
+        config = configparser.ConfigParser()
+        myGroup = "General"
+        config.add_section(myGroup)
+        myKey = "KEY"
+        myVal = "YES"
+        config[myGroup][myKey] = myVal
+        river_data = DFastMIConfigParser(config)
+        river_data._processor.register_processor(Test_DFastMIConfigParser, river_data._process_entry_value, lambda x: 1/0)
+
+        with pytest.raises(Exception) as cm:
+            river_data.read_key(Test_DFastMIConfigParser, myKey, self._reach)
+        assert str(cm.value) == f'Reading {myKey} for reach {self._reach.name} on {self._reach.parent_branch.name} returns "{myVal}". Expecting 1 values.'
+
+    def given_simple_river_config_when_read_key_with_custom_unknown_tuple_type_then_exception_thrown(self, setup_data):
+        """
+        When we want to read a configuration tuple value from a river configuration file but have not registered a type
+        or actually have an exception while parsing the value string to the tuple type we want we return an empty tuple.
+        """
+        config = configparser.ConfigParser()
+        myGroup = "General"
+        config.add_section(myGroup)
+        myKey = "KEY"
+        myVal = "YES"
+        config[myGroup][myKey] = myVal
+        river_data = DFastMIConfigParser(config)
+        river_data._processor.register_processor(Test_DFastMIConfigParser, river_data._process_tuple_entry_value, lambda x: 1/0)
```

Comment:
maybe move the parts above river_data._processor.register_processor also to setup, since that is the same in both methods, might make the difference between the methods more clear? Else move that part to a helper method?